lib/ansible/playbook/role/__init__.py

```diff
@@ -271,6 +274,64 @@ def _load_role_data(self, role_include, parent_role=None):
                 raise AnsibleParserError("The handlers/main.yml file for role '%s' must contain a list of tasks" % self._role_name,
                                          obj=handler_data, orig_exc=e)
 
+    def _prepend_validation_task(self, task_data):
+        '''Insert a role validation task if we have a role argument spec.
+
+        This method will prepend a validation task to the front of the role task
+        list to perform argument spec validation before any other tasks, if an arg spec
+        exists for the entry point. Entry point defaults to `main`.
+
+        :param task_data: List of tasks loaded from the role.
+
+        :returns: The (possibly modified) task list.
+        '''
+        if self._metadata.argument_specs:
```

Comment:
Expectations around role dependency and spec validation will probably vary widely. What we are going with currently is each role gets a validation task immediately before the role is executed. We are not special handling roles if they are dependencies and changing the validation order.

We should probably warn that role dependencies will run even if validation of the role with the deps fails. The warning message below needs some wordsmithing.

```suggestion
        if self._metadata.argument_specs:
            if self._dependencies:
                display.warning("Dependent roles will run before roles with argument specs even if validation fails.")
```

Reply:
In rewording, do not use role name(s) or any identifiers, so it only issues one warning (imagine 4 dependent roles with argspec and warning for each).

Reply:
If we wanted to add an option to disable this warning, add a new option in `lib/ansible/config/base.yml` similar to other `*_WARNING` options.

`import ansible.constants as C` then just check `C.ROLE_VALIDATION_DEP_WARNING` (or whatever you end up calling it).

Reply:
Thanks! I put this in the latest push.